### DISCUSSION TITLE
The Relation option in collections

### DISCUSSION THREAD

**Discussion by regisx001** [pocketbase/pocketbase#398] [Q&A]
can we make a relation between a record and full other collection ?

**Comment by ganigeorgiev** [comment]
The `relation` type field support only single or multiple `Record` id references.

There is no special "_relation to entire collection_" field but you can always add a `text` field and store the collection id or name (if you need server-side validations for the stored value you'll have to use the [event hooks](https://pocketbase.io/docs/event-hooks/)).

I'm curious, what is your use case for this?

**Reply by regisx001** [reply]
I working Now on a Movie application  
and I create a main collection that include all the category's (action, romantic ...), and i want to make a relation between a record and a collection that have all the movies and their info on other collection !
but now I find a solution which is create a text feild and store on it the collection name , like that i can handle the other collection from the Sdk .
 thank for your answer !

**Reply by StickOnAStick** [reply]
Another use case would be for having guest's alongside users.
A game I'm developing requires that the host of a match can be either a custom "guests" record or a signed "users" record.

**Reply by ganigeorgiev** [reply]
@StickOnAStick I don't see issue with using the `relation` field as it is for your use case. For "guests" the field value could be just empty.

Or alternatively, you can always create a system "guest" user record (_similar to GitHub's "ghost account" for deleted users_).

Another alternative if you want to have more than one "guest" could be to represent the "guests" records as regular "users" records but with a `bool` guest flag.

**Reply by StickOnAStick** [reply]
@ganigeorgiev

It's a non issue, but requires additional columns for a host and userHost alongside extra checks to ensure neither are empty at the same time. As well as having to make two database calls instead of one, expecting at least one to fail. 

The gimmicky solution is just to use a text field and store the id in there, then search both guests and users when making an api call from the front end. This removes the need for two columns but also adds issues with ensuring the text input is a valid ID, and doesn't remove the need to search both users and guests collections.

**Reply by ganigeorgiev** [reply]
@StickOnAStick I don't have full picture of your use case and I'm not sure what to recommend, but in the above suggestions it is not required to create extra calls, nor to maintain 2 collections. The main idea was that you could treat the "guests" as "users" with temp autogenerated password, aka. have only 1 "users" colleciton and a `relation` field pointing to it.

There are no plans to add "polymorphic relations" (_aka. single relation field that could have relations to more than one collection_)  as this will complicate a lot of the internals and the data representation.

If you really have to maintain 2 different collections, you can also have 2 `relation` fields (one for "guests" and one for "users") and add an API rule that require only one of the fields to be set (_I assume this is probably what you meant_), but I'd generally avoid this and will try to combine both collections in 1.

**Reply by ganigeorgiev** [reply]
Another approach could be also to inverse the relation and/or to split the common part in its own collection (eg. dedicated "hosts" collections with the shared fields and then you can have separate relations guests->hosts and users->hosts, either maintained as direct relation fields inside "guests" and "users" or as junction collections).

**Reply by StickOnAStick** [reply]
The current structure is the following 

#### Collections and fields below:
   ##### Users - Normal PB Users
   ##### Guests - Unsigned guests
    | - Current Lobby 
    | - Current Game
    | - Cookie token for ID'ing
   ##### Lobbys
    | - ... other fields
    | - Host <- Currently can only be set to guest XOR user if using relation field. 
   ##### Games
    | - ... other fields
    | - Host <- Same issue here
      
   Because of this structure, I've opted to use a text field to hold the ID of the Host within each record. Although it still requires two lookups (with one guaranteed to fail), it's a simpler approach than adding two fields _Host_ and _UserHost_ and still having to do two lookups. 
   
   I've tried another approach by creating a collection called UsersOrGuests which was structured as follows. But this approach became messy very quickly when actually integrated; hence, the text field for ID workaround
   
#### UsersOrGuests
   ##### UsersOrGuests
    | - User <----  Relation fields
    | - Guest <-╝

**Reply by ganigeorgiev** [reply]
My suggestion to merge the "users" and "guests" in a single collection remains. A "guest" could be just a random created user optionally with a "guest" bool flag. The user password could be stored temporary client-side (similar to how you are currently storing _"Cookie token for ID'ing"_). When the "guest" decide to finally "register" you can still show a user create form, but behind the scenes you can perform an update of the previously created random user.

The suggestion from https://github.com/pocketbase/pocketbase/discussions/398#discussioncomment-6731174 is also valid. For example, you can create a `hosts` collection with the shared hosts fields and an _optional_ "user" relation field for the cases where the "host" is registered user.

But again, I don't fully understand your use case and using a simple `text` field could be indeed easier.

**Reply by StickOnAStick** [reply]
The use case is to prevent guest from needing to sign in. I didn't want users of my web game to require sign in with email/pass. Making it easier for new players to play.

Because of that I was unable to merge the two together or create a guest toggle. I'm also actively trying to utilize the Auth store as much as possible for users that are signed.

**Reply by ganigeorgiev** [reply]
> The use case is to prevent guest from needing to sign in. I didn't want users of my web game to require sign in with email/pass. Making it easier for new players to play.

But that's not what I suggested in the previous comment. What I meant is that you can treat the "guest" user as regular user, but instead of creating a "guest" record, you can create a random generated "user" record **behind the scene** (the email can be configured to not be required and the only thing you'll need to provide will be a random generated password that you can store client-side). When the "guest" decide to create an account, you can "upgrade" the existing user. This flow is very similar for example when using an "anonymous user" auth in Firebase or other similar platforms.

I'll not pursue further the topic as it started to become too vague and it seems that you've already found a working workaround.
